In [14]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [15]:
# Load the given datasets
numeric = pd.read_csv("chronic_kidney_disease_numerical.csv")
categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
print(numeric)
print(categorical)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data is a set of details about different patients that can be used to predict chronic kidney disease. The features are a variety of health factors such as blood pressure, red blood cell count, and appetite (and 21 more), and the labels are "ckd" and "notckd" meaning whether the patient has chronic kidney disease or not. There is a mix of categorical and numerical data. Some categorical columns are appetite (good or poor), bacteria (present or not present), and puc cell (normal or abnormal). Some numerical columns are age, blood pressure, and red blood cell count.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [16]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print("duplicate rows in numeric: \n", numeric[numeric.duplicated()])
print("total number: ", sum(numeric.duplicated()))

# Drop these duplicate rows
numeric_nodup = numeric.drop_duplicates()
print(len(numeric_nodup))

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print("duplicate rows in categorical: \n", categorical[categorical.duplicated()])
print("total number: ", sum(categorical.duplicated()))

categorical_nodup = categorical.drop_duplicates()
print(len(categorical_nodup))

duplicate rows in numeric: 
      unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
107     872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
149     721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
171     240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
173     441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
185     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
194     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  46.0   
213     354223  64.0   70.0   97.0   27.0  0.7  145.0  4.8  13.8  49.0   
244     302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
245     431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
248     687021  52.0   80.0   99.0   25.0  0.8  135.0  3.7  15.0  52.0   
249     843449  30.0   80.0  131.0   38.0  1.0  147.0  3.8  14.1  45.0   
265     628271  65.0   90.0  270.0   39.0  2.0    NaN  NaN  12.0  36.0   
273     8

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: From looking quickly at the sorted id lists, it seems like they mostly match, however categorical has 10 more rows and no duplicates, so it must have some ids that are not in numeric. If merged while keeping all unique ids, the rows of categorical that are not in numeric will have their categorical values but NA for the numeric columns.

In [17]:
# (looking at the ids to answer the question)
numeric_ids = (numeric_nodup.sort_values(by=['unique_id']).unique_id).reset_index(drop=True)
categorical_ids = (categorical_nodup.sort_values(by=['unique_id']).unique_id).reset_index(drop=True)

print(sum(categorical_ids.duplicated()))

print(numeric_ids)
print(categorical_ids)

# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = numeric_nodup.merge(categorical_nodup, how='outer', on='unique_id')

# Print the combined dataset
print(merged)

0
0      102677
1      103570
2      104675
3      105923
4      107297
        ...  
385    995177
386    996076
387    996412
388    997629
389    997876
Name: unique_id, Length: 390, dtype: int64
0      102677
1      103570
2      104675
3      105923
4      107297
        ...  
395    995177
396    996076
397    996412
398    997629
399    997876
Name: unique_id, Length: 400, dtype: int64
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     9960

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [18]:
# Calculate the percentage of rows that contain at least one missing value
isNa = merged.isna()
rows_missing = isNa.apply(lambda x: sum(x)>0, axis=1)
num_rows_missing = sum(rows_missing)
percent_missing = num_rows_missing/len(merged)*100

# Print %
print(str(round(percent_missing, 3)) + " % rows with missing values")

# Drop these rows from the dataset
dropped_missing = merged[rows_missing[merged.index] == False]

# Print the Dataset
print(dropped_missing)

61.75 % rows with missing values
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su     rbc        pc         pcc          ba  htn  \

## Part 3.4: Sort the dataset according to the Labels

In [19]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
target_sorted = dropped_missing.sort_values(by=['Target']).reset_index(drop=True)

# Print the dataset
print(target_sorted)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [20]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
for column in categorical_columns:
    codes, uniques = pd.factorize(target_sorted[column])
    for unique in uniques:
        target_sorted[column + "_" + unique] = list(map(lambda x: int(uniques[x]==unique), codes))

cat_encoded = target_sorted.drop(columns=categorical_columns)


# Print the dataset
print(cat_encoded)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_normal  rbc_abnormal  pc_abnormal  \
0     3800.0   2.1  3.0  0.0           1

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

If there are 5 separate columns for one feature, it would be treated as 5 different features by modeling. So, it would make the dimensions you have to work with much higher. It might still be able to make good predictions, but I'm guessing it would use more resources.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [21]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

remove_outliers = cat_encoded.copy()

for column in numerical_columns:
    std = remove_outliers[column].std()
    mean = remove_outliers[column].mean()
    remove_outliers = remove_outliers[abs(remove_outliers[column]-mean) < 3*std]

remove_outliers = remove_outliers.reset_index(drop=True)
    
# Print the dataset
print(remove_outliers)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
128     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0   
129     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0   
130     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0   
131     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0   
132     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_normal  rbc_abnormal  pc_abnormal  \
0     9200.0   3.2  2.0  0.0           0

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [22]:
# Normalize the all Numerical Attributes in the dataset.
normalized = remove_outliers.copy()

for column in normalized:
    normalized[column] = (normalized[column] - normalized[column].min()) / (normalized[column].max() - normalized[column].min())   

# Print the dataset
print(normalized)

     unique_id       age    bp       bgr        bu        sc       sod  \
0     0.264544  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1     0.325059  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2     0.371049  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3     0.411888  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4     0.412782  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
128   0.425405  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
129   0.106438  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
130   0.420566  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
131   0.135054  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
132   1.000000  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv      wbcc      rbcc   al   su  rbc_normal  \
0    0.793103  0.000000  0.03225

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

The unique_id column is not appropriate. It is simply a random number used to identify a patient. It was helpful when combining the datasets, but it should not be used for modeling and predictions since it is not a health factor and should not be used to predict chronic kidney disease.

Answer:

In [23]:
#Remove that column
final_data = normalized.drop(columns=['unique_id'])

# Print the dataset
print(final_data)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
1    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
2    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
3    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
4    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
128  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
129  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
130  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
131  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
132  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc      rbcc   al   su  rbc_normal  \
0    0.000000  0.032258  0.395161  0.057143  0.5  0.0 

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [24]:
# Export the dataframe to a new csv file
final_data.to_csv("final_data_results.csv", index=False)

# Export the dataframe to a new json file
final_data.to_json("final_data_results.json", orient="split")


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow


| age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_normal | rbc_abnormal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_yes | htn_no | dm_no | dm_yes | cad_no | cad_yes | appet_poor | appet_good | pe_yes | pe_no | ane_yes | ane_no | Target_ckd | Target_notckd |
|------------|--------|-----------|----------|----------|-----------|----------|-------------|-----------|---------------|-------------|---------|--------|------------|---------------|-------------|-----------|----------------|--------------|----------------|------------|---------|--------|-------|--------|--------|---------|------------|-------------|--------|-------|---------|--------|-------------|----------------|
| 0.743243 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.032258 | 0.395161 | 0.057143 | 0.50 | 0.00 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 |
| 0.770270 | 1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667 | 0.206897 | 0.524752 | 0.548387 | 0.443548 | 0.342857 | 0.50 | 0.50 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 |
| 0.905405 | 0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000 | 1.000000 | 0.277228 | 0.322581 | 0.233871 | 0.371429 | 0.50 | 0.00 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 |
| 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.000000 | 0.653226 | 0.257143 | 1.00 | 0.00 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 |
| 0.783784 | 1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.586207 | 0.019802 | 0.096774 | 0.258065 | 0.114286 | 1.00 | 0.50 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 |
| 0.729730 | 0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.064516 | 0.879032 | 0.000000 | 0.75 | 0.25 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 |
| 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548 | 0.169355 | 0.114286 | 0.50 | 0.00 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 |
| 0.851351 | 0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.344828 | 0.168317 | 0.161290 | 0.580645 | 0.085714 | 1.00 | 0.75 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 |
| 0.540541 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.161290 | 0.830645 | 0.057143 | 0.25 | 0.00 | 1.0 | 0.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 0.0

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Using the GPT was not too helpful for this task. It would purposely abbreviate rows or simply not finish generating all the rows of the table, so here it only shows 9. In general, ways to mitigate bad responses could be by having it write out what it is doing at different steps so it stays on track, or giving it the data in a clean format, like uploading the csv file directly, since pasting output from somewhere else can have odd spacing or symbols.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [25]:
# Code to rename all the columns in the dataset
renamed_data = final_data.copy()

# Dictionary to map current column names to real names
real_names = {
    'age': 'Age (year)',
    'bp': 'Blood Pressure (mm/Hg)',
    'bgr': 'Blood Glucose Random (mg/dl)',
    'bu': 'Blood Urea (mg/dl)',
    'sc': 'Serum Creatinine (mg/dl)',
    'sod': 'Sodium (mEq/L)',
    'pot': 'Potassium (mEq/L)',
    'hemo': 'Hemoglobin (gms)',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count (cells/cmm)',
    'rbcc': 'Red Blood Cell Count (millions/cmm)',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_normal': 'Red Blood Cells (Normal)',
    'rbc_abnormal': 'Red Blood Cells (Abnormal)',
    'pc_abnormal': 'Pus Cells (Abnormal)',
    'pc_normal': 'Pus Cells (Normal)',
    'pcc_notpresent': 'Pus Cell Clumps (Not Present)',
    'pcc_present': 'Pus Cell Clumps (Present)',
    'ba_notpresent': 'Bacteria (Not Present)',
    'ba_present': 'Bacteria (Present)',
    'htn_yes': 'Hypertension (Yes)',
    'htn_no': 'Hypertension (No)',
    'dm_no': 'Diabetes Mellitus (No)',
    'dm_yes': 'Diabetes Mellitus (Yes)',
    'cad_no': 'Coronary Artery Disease (No)',
    'cad_yes': 'Coronary Artery Disease (Yes)',
    'appet_poor': 'Appetite (Poor)',
    'appet_good': 'Appetite (Good)',
    'pe_yes': 'Pedal Edema (Yes)',
    'pe_no': 'Pedal Edema (No)',
    'ane_yes': 'Anemia (Yes)',
    'ane_no': 'Anemia (No)',
    'Target_ckd': 'CKD or Not CKD',
    'Target_notckd': 'Not CKD'
}

# Rename the columns
renamed_data.rename(columns=real_names, inplace=True)

# Display the updated DataFrame
print(renamed_data)

     Age (year)  Blood Pressure (mm/Hg)  Blood Glucose Random (mg/dl)  \
0      0.743243                    0.50                      0.442060   
1      0.770270                    1.00                      0.901288   
2      0.905405                    0.50                      0.785408   
3      0.202703                    0.75                      0.158798   
4      0.783784                    1.00                      0.399142   
..          ...                     ...                           ...   
128    0.189189                    0.25                      0.227468   
129    0.216216                    0.00                      0.115880   
130    0.513514                    0.00                      0.107296   
131    0.702703                    0.50                      0.128755   
132    0.689189                    0.00                      0.150215   

     Blood Urea (mg/dl)  Serum Creatinine (mg/dl)  Sodium (mEq/L)  \
0              0.901961                  0.432099     

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [26]:
# Converted SQL to Pandas code

# Group by the 'Target' column and count occurrences
result = target_sorted.groupby('Target').size().reset_index(name='count')


# Display the result
print(result)


   Target  count
0     ckd     42
1  notckd    111


*Describe what the above code does here*

It groups the code by the Target values, counts the number of rows with each value, resets the indices to turn it back into a dataframe, and names the count column "count". So, in the end, it displays the counts of how many patients had chronic kidney disease and how many did not.